include/multipass/async_periodic_task.h

```diff
@@ -60,16 +66,41 @@ class AsyncPeriodicTask
 
         // TODO, remove the launch_msg parameter once we have better class separation.
         mpl::log(mpl::Level::debug, "async task", std::string(launch_msg));
-        future = std::async(std::launch::async, std::forward<Callable>(func), std::forward<Args>(args)...);
+        future = QtConcurrent::run(std::forward<Callable>(func), std::forward<Args>(args)...);
+
+        auto event_handler_on_sucess_and_failure = [retry_start_delay_time, this]() -> void {
+            try
+            {
+                // rethrow exception
+                future.waitForFinished();
+                // success case, we reset the time interval for timer and the reset the retry delay time value
+                timer.start(default_delay_time);
+                retry_current_delay_time = retry_start_delay_time;
+            }
+            catch (const multipass::DownloadException& e)
```

Comment:
If `async_periodic_task` is meant to be generic, shouldn't we instead catch something a bit more generic so that this isn't tied to only doing download tasks?

Reply:
Excellent question and a nice observation, but I think retry should only act on the DownloadException and not act on other types of exceptions. 

On the other hand, `async_periodic_task` is a generic utility class that handles both download and non-download types of tasks; in the case of non-download type classes, we may not need this retry mechanism. Then maybe it is better to decouple the retry part code from the class. That will be some work. 

For now, there are only async periodic download tasks that occur, so changing this class name to `AsyncPeriodicDownloadTask` and making it only launches download tasks is also a good solution.

Reply:
Just to add my two cents, you both have good points and I think renaming is a good compromise for the time being.

Reply:
I'm fine with the renaming as well.

I think in the future if we needed to create an `AsyncPeriodicTask` without a retry timer we would add another constructor with `retry_start_delay_time` omitted that simply uses `INT_MAX`, or something of the like, as a default delay.

Reply:
@georgeliao The filename is still unchanged

Reply:
Nice catch, overlooked that,

Reply:
Ok, I think this is good for the time being, however I would like to see it become more generic in the future. @ricab, if you are fine with this, then I think it can go in.

Reply:
Thanks @sharder996, I did my review already so feel free to merge when you approve.